brave/src/main/java/brave/secondary_sampling/SecondarySampler.java

```diff
@@ -0,0 +1,45 @@
+/*
+ * Copyright 2019 The OpenZipkin Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License
+ * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+ * or implied. See the License for the specific language governing permissions and limitations under
+ * the License.
+ */
+package brave.secondary_sampling;
+
+import brave.http.HttpServerRequest;
+import brave.propagation.TraceContext;
+
+public interface SecondarySampler {
+  /**
+   * Returns a trigger function that decides if this request is sampled for the given {@code
+   * samplingKey}.
+   *
+   * <p>This is invoked during {@link TraceContext.Extractor#extract(Object)}. It is usually the
+   * case that the carrier parameter is a request. For example, in http middleware, it will be
+   * {@link HttpServerRequest}. However, in other RPC middleware it may be a different type.
+   *
+   * <p>For example, to only handle http requests, implement like so:
+   * <pre>{@code
+   *     if (!(carrier instanceof HttpServerRequest)) return false;
```

Comment:
Like how secondary sampling can be started by matching a grpc method name or a http endpoint, should it be possible to do the same for B3 sampling also? Which means a http request/grpc call should be matched in one place and rule applied for both b3 sampling or secondary sampling?

For example, imagine the following config to start a trace;
```
  samplingpolicies:
    - "k=sayHello50pct;m=SayHello;p=0.5"
    - "m=SayHi;p=1.0"
```

Setting this on a `HelloService` to enable the following sampling;
- Start a secondary sampling to trace 50% of `SayHello` grpc method
- Start b3 sampling to trace all `SayHi` grpc method

This means should `SecondarySamplingState` used for only matching service participation and updating ttl but not starting a new trace?

High level ask is, should starting a trace by setting b3 or sampling_key done in the one place so applying the matching rule can be done in one place.

Fwiw, I was looking at `brave.http.HttpRuleSampler` and wondered should that be reused for secondary sampling as well.

Reply:
To merge primary and secondary configuration together is possible, but the parameters are different of course. Remember, SecondarySampler is below the http abstraction.. it is a foundation for http and also normal rpc or messaging.

HttpRuleSampler can only work inside the http abstraction.

Anyway you can open an issue about an example of shared controller of primary and secondary sampling. That work is separate from this refactoring, though happy this work jogged the idea.

Reply:
PS It is neither my intent nor in range of available bandwidth to create a config management system for trace sampling policy. When you raise the issue I can try to help show you how to hook things together such that you can make one.

Reply:
I'm readying for my last flight, but here's the plumbing needed. I can change the javadoc accordingly. https://github.com/openzipkin/brave/pull/994

Note: one difference is that the HttpSampler returns nullable result as a signal to use probabilistic sampling (ex not based on the request).

We can either coerce null to false here, or think about if there's a use case to use the tracer's default sampler when there's no match. If we did, we'd need to wire that into the parameters to create the secondary sampling instance.

Reply:
I made the issue for you. please do this on your own next time as it is good behavior https://github.com/openzipkin-contrib/zipkin-secondary-sampling/issues/8